### ISSUE TITLE
AIP-38 | Add colorblind friendly color schemes

### ISSUE THREAD

**Issue by bbovenzi** [apache/airflow#43054]
We are starting to add status colors to the new UI. Before we build out too much we should make sure to think about a11y from the get go. A lot of status colors are hard for people (particularly red/green).

- Update our theme to always use semantic names instead of colors (ex: "error" instead of "red", "success" instead of "green)

- Decide to use a [colorblind friendly color palette for everyone](https://personal.sron.nl/~pault/), or add a user setting to switch between common semantic colors and a friendlier palette

**Comment by potiuk** [comment]
❤️

**Comment by bbovenzi** [comment]
We have 12 state colors right now, which is a lot, and leaves us with three options:

1. Switch our state colors to an imperfect colorblind friendly palette for everyone. No need for settings but a lot of colors will really change: https://mk.bcgsc.ca/colorblind/palettes.mhtml#12-color-palette-for-colorbliness

Ex: ![Image](https://github.com/user-attachments/assets/1e45877a-febe-4f78-8cd0-db4c5b898faa)

2. Add a color-blind friendly mode to allow users to opt-in to a more accessible mode

3. Use a better color blind friendly palette, but with fewer colors and the need to consolidate task states



In all scenarios, we should always try to depict state through other means than simply color.

**Comment by bbovenzi** [comment]
Possible groupings with the same color but some indication that they're separate:

Pre run: queued, schedule, deferred
Re do: restarting, up_for_retry, up_for_reschedule
Nothing: no_status, skipped, removed, upstream_failed
Failed
Success
Running

**Comment by ashb** [comment]
I think queued and scheduled are distinct enough to need different colors.

And skipped /upstream_failedcould be the same colour, but I don't think no_status/null should be in the same grouping. One has never been, the other ran and failed. Removed I could se go in either side.

**Comment by shahar1** [comment]
> 3\. Use a better color blind friendly palette, but with fewer colors and the need to consolidate task states.
>     With dag versioning, will we have Removed or just None or Skipped?
>     Or can we move to a status/sub status model

In general - great idea and kudos for making the UI more accessible for those in need!

My two cents regarding your suggestions:
1. Not too bad, but some of the colors in the pallette are too similar:
  a. restarting - skipping
  b. deferred - removed
2. In any case I'd make this transition optional for now (long term - maybe even user customized?).
3. I think that I'd go with the first option rather than having less and too similar colors.
If we go in the direction of less colors, maybe we could utilize other features like opacity/fill/shape?

+ Don't forget the dark mode :)

**Comment by bbovenzi** [comment]
1. With some user testing, the 12 color palette I linked didn't help much.

2. I am definitely moving forward with a version of this. Looking at a hue rotation that let's a user shift the whole UI as they see fit: https://stackoverflow.com/questions/46553023/use-slider-to-change-image-css-filter-with-javascript/46553057#46553057

3. I need to figure out which other changes can work well with colors. I definitely like the idea of visually separating tasks that are in pending or final states. Pending can be a circle while final is more square shaped or we can animate a gradient bg for pending states.

**Comment by potiuk** [comment]
FYI Animation is also against guidelines for accessibility - we had issue about it https://github.com/apache/airflow/issues/34019

**Comment by potiuk** [comment]
We could also let the user pick their own colors for each state.

**Comment by bbovenzi** [comment]
Yes, we do allow `STATE_COLORS` to be edited via settings but on a deployment, not a user basis. I don't think that should exist in the browser's localstorage either. It would be great to create actual user settings we save on the db which can also include other requests like "favorite_dags".

**Comment by vmatt** [comment]
hey @bbovenzi and the rest of the community!

What do you thing of introducing symbols instead of using colors? I've [created a userscript](https://greasyfork.org/en/scripts/518865-airflow-task-instance-status-enhancer) which does just that, but it's kinda hacky, and interferes with React's Virtual DOM. It would be way better to implement something similar natively.

**Comment by jscheffl** [comment]
> hey @bbovenzi and the rest of the community!
> 
> What do you thing of introducing symbols instead of using colors? I've [created a userscript](https://greasyfork.org/en/scripts/518865-airflow-task-instance-status-enhancer) which does just that, but it's kinda hacky, and interferes with React's Virtual DOM. It would be way better to implement something similar natively.
> 
> With this change, the default color scheme would remain, but users could switch to symbols / tweak it in their personal user settings.
> 

I like the idea with icons - we just need to consider that while the enlarged mugshot look nice - especially with the details in the icons - on most screens (non retina) this will be just a few pixels. So symbols can help BUT we need to make them also visually distinguishable in todays 16 x 16 pixel raster (with the option to have more details in retina displays)

![image](https://github.com/user-attachments/assets/86900c61-905d-47ba-9cb6-4173614d13e9)

So if we go for symbols they would need to be very reduced in detail and colors and need to be distinguishable being max 14 x 14 pixels (assuming 2 pixels border to the next icon)

**Comment by potiuk** [comment]
Yep. I m also for symbols - but they should be done RIGHT.

**Comment by bbovenzi** [comment]
In case anyone missed it. We now have icons for each task state. This is implemented everywhere in the new UI except for the grid view for now.

<img width="1175" alt="Image" src="https://github.com/user-attachments/assets/3ca106d1-e0e1-4db5-9781-0476b8cb99ce" />
<img width="1190" alt="Image" src="https://github.com/user-attachments/assets/c61ce9a7-d180-42e8-a928-77a9e9f9e9e8" />
<img width="1152" alt="Image" src="https://github.com/user-attachments/assets/19e1d6df-44f0-43bc-9d3c-e6af487be0d7" />

**Comment by FrankPortman** [comment]
Why was this closed as completed @bbovenzi? Because of icons? AFAICT `STATE_COLORS` support and documentation has been quietly removed in Airflow3.

**Comment by bbovenzi** [comment]
Yes, in Airflow 3 we retired a lot of old settings when rewriting the frontend. We ran into a lot of issues of having the config having control UI elements like colors, especially for light+dark mode and various shades. We are looking to bring it back in the form of a larger theme config to allow better customization.
